File: references/betterlectio/ios/BetterLectio/SupabaseSchoolService.swift
//
//  SupabaseSchoolService.swift
//  BetterLectio
//

import Foundation
import Supabase

final class SupabaseSchoolService {
    static let shared = SupabaseSchoolService()

    private let manager: SupabaseManager

    init(manager: SupabaseManager = .shared) {
        self.manager = manager
    }

    func fetchAllSchools() async throws -> [School] {
        guard let client = manager.client else { return [] }

        let rows: [SchoolRow] = try await client
            .from("schools")
            .select("id,name")
            .order("name", ascending: true)
            .execute()
            .value

        return rows.map { School(id: Int($0.id), name: $0.name) }
    }
}

private struct SchoolRow: Decodable {
    let id: Int64
    let name: String
}

File: references/betterlectio/android/app/src/main/java/dk/betterlectio/android/feature/supabase/SupabaseSchoolService.kt
package dk.betterlectio.android.feature.supabase

import dk.betterlectio.android.core.model.School
import io.github.jan.supabase.postgrest.from
import io.github.jan.supabase.postgrest.query.Order
import kotlinx.serialization.Serializable
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Schools catalog from Supabase `schools` table (iOS: `SupabaseSchoolService`).
 */
@Singleton
class SupabaseSchoolService @Inject constructor(
    private val manager: SupabaseManager,
) {
    suspend fun fetchAllSchools(): List<School> {
        val client = manager.client ?: return emptyList()
        // Schools are public/anon-readable; no need to wait for user session.
        return try {
            val rows = client.from("schools")
                .select {
                    filter {
                        // select all rows; order applied below
                    }
                    order("name", Order.ASCENDING)
                }
                .decodeList<SchoolRow>()
            rows.map { School(id = it.id.toInt(), name = it.name) }
        } catch (e: Exception) {
            Timber.w(e, "Supabase schools fetch failed")
            emptyList()
        }
    }

    @Serializable
    private data class SchoolRow(
        val id: Long,
        val name: String,
    )
}
